core/src/main/java/org/infinispan/interceptors/MarshalledValueInterceptor.java

```diff
@@ -175,6 +169,29 @@ private Object visitDataReadCommand(InvocationContext ctx, AbstractDataCommand c
    }
 
    @Override
+   public Object visitGetAllCommand(InvocationContext ctx, GetAllCommand command) throws Throwable {
+      if (wrapKeys) {
+         Set<Object> marshalledKeys = new HashSet<>();
+         for (Object key : command.getKeys()) {
+            if (!isTypeExcluded(key.getClass())) {
+               MarshalledValue mv = createMarshalledValue(key, ctx);
+               marshalledKeys.add(mv);
+            } else {
+               marshalledKeys.add(key);
+            }
+         }
+         command.setKeys(marshalledKeys);
+      }
+      Map<Object, Object> map = (Map<Object, Object>) invokeNextInterceptor(ctx, command);
+      Map<Object, Object> unmarshalled = command.createMap();
+      for (Map.Entry<Object, Object> entry : map.entrySet()) {
+         // TODO: how does this apply to CacheEntries if command.isReturnEntries()?
```

Comment:
`cache.getCacheEntry(k)` doesn't seem to unmarshal the value, but I'm not sure that's intentional. It would probably be good to keep the value marshalled on remote nodes and only unmarshal it on the originator.